File: web/src/features/nutrition/api/nutritionApi.ts
// =====================================================================
// MiCoach — Cliente REST del módulo nutrition (/api/v1/nutrition).
// Paridad con mobile/lib/features/nutrition/infrastructure/nutrition_api.dart.
// =====================================================================
import { apiClient } from '@/core/api/client';
import type {
  DailyIntakeEntry,
  Ingredient,
  MealPlan,
  MealPlanDraft,
  Recipe,
  RecipeFilter,
  ShoppingList,
  ShoppingListItem,
  Substitution,
} from '../domain/nutritionTypes';
import { mealPlanDraftToPayload } from '../domain/nutritionTypes';

export const nutritionApi = {
  listIngredients: (params: { category?: string; search?: string }) =>
    apiClient.get<Ingredient[]>('/nutrition/ingredients', { params }).then((r) => r.data),

  listRecipes: (filter: RecipeFilter) =>
    apiClient
      .get<Recipe[]>('/nutrition/recipes', {
        params: { mealCategory: filter.mealCategory, difficulty: filter.difficulty, search: filter.search || undefined },
      })
      .then((r) => r.data),

  getRecipe: (id: number) => apiClient.get<Recipe>(`/nutrition/recipes/${id}`).then((r) => r.data),

  listSubstitutions: (ingredientId: number) =>
    apiClient.get<Substitution[]>(`/nutrition/ingredients/${ingredientId}/substitutions`).then((r) => r.data),

  generateSubstitution: (ingredientId: number, reason: string, notes?: string) =>
    apiClient
      .post<Substitution>(`/nutrition/ingredients/${ingredientId}/substitutions/generate`, { reason, notes })
      .then((r) => r.data),

  listMealPlans: () => apiClient.get<MealPlan[]>('/nutrition/meal-plans').then((r) => r.data),

  getMealPlan: (id: number) => apiClient.get<MealPlan>(`/nutrition/meal-plans/${id}`).then((r) => r.data),

  createMealPlan: (draft: MealPlanDraft) =>
    apiClient.post<MealPlan>('/nutrition/meal-plans', mealPlanDraftToPayload(draft)).then((r) => r.data),

  generateMealPlan: (goal: string) =>
    // Igual que workoutApi.generateWorkout: la IA local (Ollama, CPU) tarda bastante más
    // que una llamada normal.
    apiClient.post<MealPlan>('/nutrition/meal-plans/generate', { goal }, { timeout: 180000 }).then((r) => r.data),

  updateMealPlan: (id: number, draft: MealPlanDraft) =>
    apiClient.put<MealPlan>(`/nutrition/meal-plans/${id}`, mealPlanDraftToPayload(draft)).then((r) => r.data),

  adjustMealPlanCalories: (id: number) =>
    apiClient.post<MealPlan>(`/nutrition/meal-plans/${id}/adjust-calories`, undefined, { timeout: 180000 }).then((r) => r.data),

  deleteMealPlan: (id: number) => apiClient.delete(`/nutrition/meal-plans/${id}`),

  listIntake: (date?: string) =>
    apiClient.get<DailyIntakeEntry[]>('/nutrition/intake', { params: { date } }).then((r) => r.data),

  logIntake: (body: {
    recipeId?: number | null;
    foodDate: string;
    mealType: string;
    amount?: number | null;
    calories?: number | null;
    proteinG?: number | null;
    carbsG?: number | null;
    fatG?: number | null;
  }) => apiClient.post<DailyIntakeEntry>('/nutrition/intake', body).then((r) => r.data),

  deleteIntake: (id: number) => apiClient.delete(`/nutrition/intake/${id}`),

  listShoppingLists: () => apiClient.get<ShoppingList[]>('/nutrition/shopping-lists').then((r) => r.data),

  getShoppingList: (id: number) => apiClient.get<ShoppingList>(`/nutrition/shopping-lists/${id}`).then((r) => r.data),

  createShoppingList: (body: { name?: string; weekStart?: string | null }) =>
    apiClient.post<ShoppingList>('/nutrition/shopping-lists', body).then((r) => r.data),

  deleteShoppingList: (id: number) => apiClient.delete(`/nutrition/shopping-lists/${id}`),

  addShoppingListItem: (
    shoppingListId: number,
    body: { ingredientId?: number | null; itemName?: string | null; amount?: number | null; unit?: string | null; category?: string | null },
  ) => apiClient.post<ShoppingListItem>(`/nutrition/shopping-lists/${shoppingListId}/items`, body).then((r) => r.data),

  setItemChecked: (shoppingListId: number, itemId: number, checked: boolean) =>
    apiClient
      .put<ShoppingListItem>(`/nutrition/shopping-lists/${shoppingListId}/items/${itemId}`, { checked })
      .then((r) => r.data),

  deleteShoppingListItem: (shoppingListId: number, itemId: number) =>
    apiClient.delete(`/nutrition/shopping-lists/${shoppingListId}/items/${itemId}`),
};

File: mobile/lib/features/nutrition/infrastructure/nutrition_api.dart
// =====================================================================
// MiCoach — Cliente REST del módulo nutrition (/api/v1/nutrition).
// =====================================================================
import 'package:dio/dio.dart';

import '../domain/daily_intake.dart';
import '../domain/ingredient.dart';
import '../domain/meal_plan.dart';
import '../domain/meal_plan_draft.dart';
import '../domain/recipe.dart';
import '../domain/recipe_filter.dart';
import '../domain/shopping_list.dart';

class NutritionApi {
  final Dio _dio;

  NutritionApi(this._dio);

  // ------------------------- Catálogo -------------------------

  Future<List<Ingredient>> listIngredients({String? category, String? search}) async {
    final response = await _dio.get('/nutrition/ingredients',
        queryParameters: {if (category != null) 'category': category, if (search != null) 'search': search});
    return (response.data as List<dynamic>).map((e) => Ingredient.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<List<Recipe>> listRecipes(RecipeFilter filter) async {
    final response = await _dio.get('/nutrition/recipes', queryParameters: filter.toQueryParams());
    return (response.data as List<dynamic>).map((e) => Recipe.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<Recipe> getRecipe(int id) async {
    final response = await _dio.get('/nutrition/recipes/$id');
    return Recipe.fromJson(response.data as Map<String, dynamic>);
  }

  // ------------------------- Planes de alimentación -------------------------

  Future<List<MealPlan>> listMealPlans() async {
    final response = await _dio.get('/nutrition/meal-plans');
    return (response.data as List<dynamic>).map((e) => MealPlan.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<MealPlan> getMealPlan(int id) async {
    final response = await _dio.get('/nutrition/meal-plans/$id');
    return MealPlan.fromJson(response.data as Map<String, dynamic>);
  }

  Future<MealPlan> createMealPlan(MealPlanDraft draft) async {
    final response = await _dio.post('/nutrition/meal-plans', data: draft.toJson());
    return MealPlan.fromJson(response.data as Map<String, dynamic>);
  }

  Future<MealPlan> updateMealPlan(int id, MealPlanDraft draft) async {
    final response = await _dio.put('/nutrition/meal-plans/$id', data: draft.toJson());
    return MealPlan.fromJson(response.data as Map<String, dynamic>);
  }

  Future<void> deleteMealPlan(int id) => _dio.delete('/nutrition/meal-plans/$id');

  /// Genera un plan de alimentación con IA a partir de un pedido en lenguaje natural.
  /// Puede tardar hasta un par de minutos (proveedor local por CPU) — timeout largo.
  Future<MealPlan> generateMealPlan(String goal) async {
    final response = await _dio.post(
      '/nutrition/meal-plans/generate',
      data: {'goal': goal},
      options: Options(sendTimeout: const Duration(seconds: 10), receiveTimeout: const Duration(seconds: 180)),
    );
    return MealPlan.fromJson(response.data as Map<String, dynamic>);
  }

  // ------------------------- Diario alimentario -------------------------

  Future<List<DailyIntakeEntry>> listIntake({DateTime? date}) async {
    final response = await _dio.get('/nutrition/intake', queryParameters: {
      if (date != null) 'date': date.toIso8601String().split('T').first,
    });
    return (response.data as List<dynamic>)
        .map((e) => DailyIntakeEntry.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<DailyIntakeEntry> logIntake({
    int? recipeId,
    required DateTime foodDate,
    required String mealType,
    double? amount,
    double? calories,
    double? proteinG,
    double? carbsG,
    double? fatG,
  }) async {
    final response = await _dio.post('/nutrition/intake', data: {
      'recipeId': recipeId,
      'foodDate': foodDate.toIso8601String().split('T').first,
      'mealType': mealType,
      'amount': amount,
      'calories': calories,
      'proteinG': proteinG,
      'carbsG': carbsG,
      'fatG': fatG,
    });
    return DailyIntakeEntry.fromJson(response.data as Map<String, dynamic>);
  }

  Future<void> deleteIntake(int id) => _dio.delete('/nutrition/intake/$id');

  // ------------------------- Listas de compra -------------------------

  Future<List<ShoppingList>> listShoppingLists() async {
    final response = await _dio.get('/nutrition/shopping-lists');
    return (response.data as List<dynamic>).map((e) => ShoppingList.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<ShoppingList> getShoppingList(int id) async {
    final response = await _dio.get('/nutrition/shopping-lists/$id');
    return ShoppingList.fromJson(response.data as Map<String, dynamic>);
  }

  Future<ShoppingList> createShoppingList({String? name, DateTime? weekStart}) async {
    final response = await _dio.post('/nutrition/shopping-lists', data: {
      'name': name,
      'weekStart': weekStart?.toIso8601String().split('T').first,
    });
    return ShoppingList.fromJson(response.data as Map<String, dynamic>);
  }

  Future<void> deleteShoppingList(int id) => _dio.delete('/nutrition/shopping-lists/$id');

  Future<ShoppingListItem> addShoppingListItem(
    int shoppingListId, {
    int? ingredientId,
    String? itemName,
    double? amount,
    String? unit,
    String? category,
  }) async {
    final response = await _dio.post('/nutrition/shopping-lists/$shoppingListId/items', data: {
      'ingredientId': ingredientId,
      'itemName': itemName,
      'amount': amount,
      'unit': unit,
      'category': category,
    });
    return ShoppingListItem.fromJson(response.data as Map<String, dynamic>);
  }

  Future<ShoppingListItem> setItemChecked(int shoppingListId, int itemId, bool checked) async {
    final response =
        await _dio.put('/nutrition/shopping-lists/$shoppingListId/items/$itemId', data: {'checked': checked});
    return ShoppingListItem.fromJson(response.data as Map<String, dynamic>);
  }

  Future<void> deleteShoppingListItem(int shoppingListId, int itemId) =>
      _dio.delete('/nutrition/shopping-lists/$shoppingListId/items/$itemId');
}
